src/metabase/db/custom_migrations.clj

```diff
@@ -1759,3 +1760,37 @@
 ;; [[metabase.notification.task.send/init-send-notification-triggers!]]
 (define-migration MigrateAlertToNotification
   (pulse-to-notification/migrate-alerts!))
+
+;; copied from metabase.models.serialization
+(defn- raw-hash
+  [target]
+  (format "%08x" (hash target)))
+
+;; copied from metabase.models.serialization
+(defn- generate-nano-id
+  [seed-str]
+  (let [seed (Long/parseLong seed-str 16)
+        rnd  (Random. seed)
+        gen  (nano-id/custom
+              "_-0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+              21
+              (fn [len]
+                (let [ba (byte-array len)]
+                  (.nextBytes rnd ba)
+                  ba)))]
+    (gen)))
+
+(defn- generate-database-entity-id
+  "Generates a hash-based `entity_id` for a database."
+  [{:keys [name engine] :as _database}]
+  (-> [name engine] raw-hash generate-nano-id))
+
+(define-migration BackfillDatabaseEntityIds
```

Comment:
We can only do that because we know that we have less than 1k databases per instance.